### PR TITLE
Support for outer-scope variables in nested function (GraphProto)

### PR DESCRIPTION
**Pull request by gramalingam**
Add support for GraphProto attributes that reference outer-scope variables.

* In eager-mode, when invoking ORT for an op with GraphProto attributes, we lookup the values of outer-scope variables referenced by the graph, and add them as extra graph-inputs to pass them on to ORT.
* At script-time, check that the outer-scope variables are not redefined in between the function-definition site, and the place where the function is used as a GraphProto attribute. Such redefinitions are mapped to distinct variables in ONNX due to its SSA requirement.
* Update the analyses to account for variables used within nested GraphProto (function).
*

### CODE REVIEW

File: onnxscript/eager_mode_evaluator.py [hunk 1]
```diff
@@ -92,9 +92,10 @@ def ort_to_os_value(v):
     raise TypeError(f"Unexpected ORT value type {type(v)}.")
 
 
-def call_ort(schema, *args, **kwargs):
+def call_ort(schema, args, kwargs, implicit_args={}):
```
Comment (xadupre) [inline onnxscript/eager_mode_evaluator.py:95]: implicit_args=None (empty dictionary is not recommended for default value)

File: onnxscript/converter.py [hunk 2]
```diff
@@ -406,6 +401,20 @@ def translate_attr(self, attr_name, expr):
...
+            if isinstance(val, irbuilder.Function):
+                # Check that outer-scope variables referenced by function have same value
+                # at function-definition site and use-as-attribute site, to avoid errors.
+                for pyvar, previous in val.outer_scope_variables:
+                    current = self.lookup(pyvar, debuginfo.DebugInfo(expr, self))
+                    if current.value != previous.value:
+                        self.fail(
+                            expr,
+                            f"Outer scope variable {pyvar} referenced by function "
+                            f"{expr.id} modified."
```
Comment (xadupre) [inline onnxscript/converter.py:413]: ```{expr.id!r}``

File: onnxscript/converter.py [hunk 3]
```diff
@@ -406,6 +401,20 @@ def translate_attr(self, attr_name, expr):
             val = self.lookup(expr.id, debuginfo.DebugInfo(expr, self))
             if isinstance(val, values.AttrRef):
                 return self.ir_builder.attr_ref(attr_name, val.value, val.typeinfo)
+            if isinstance(val, irbuilder.Function):
+                # Check that outer-scope variables referenced by function have same value
+                # at function-definition site and use-as-attribute site, to avoid errors.
+                for pyvar, previous in val.outer_scope_variables:
+                    current = self.lookup(pyvar, debuginfo.DebugInfo(expr, self))
+                    if current.value != previous.value:
```
Comment (xadupre) [inline onnxscript/converter.py:409]: Is there a unit test to check this exception?

Comment (gramalingam) [reply onnxscript/converter.py:409]: Yes, see `test_outer_scope_redefinition`.

File: onnxscript/eager_mode_evaluator.py (hunk 1 above)
Comment (gramalingam) [reply onnxscript/eager_mode_evaluator.py]: Done

File: onnxscript/values.py [hunk 4]
```diff
@@ -104,23 +106,38 @@ def get_schema(self):
...
-                        "attribute. Did you forget to decorate it with @graph?"
-                    )
-        return kwargs
+            if isinstance(v, OnnxClosure):
+                kwargs[k] = v.function_ir.to_graph_proto()
+                for pyvar, onnxvar in v.function_ir.outer_scope_variables:
+                    closure[onnxvar.value] = v.frame.f_locals[pyvar]
+            elif callable(v):
+                raise ValueError(
+                    f"Error: function-valued attribute {v.__name__} has no graph_proto"
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/values.py]: Done

File: onnxscript/converter.py (hunk 2 above)
Comment (gramalingam) [reply onnxscript/converter.py]: Done

**Comment by gramalingam** [comment]
I am merging this in, to facilitate other changes. If there is any concern, please let me know, I can fix them. Thanks

**Pull request merged**